<a href="https://colab.research.google.com/github/calarconf/AlgorithmsUN2024I/blob/main/Copia_de_Lab4.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
!cat /proc/cpuinfo

processor	: 0
vendor_id	: GenuineIntel
cpu family	: 6
model		: 79
model name	: Intel(R) Xeon(R) CPU @ 2.20GHz
stepping	: 0
microcode	: 0xffffffff
cpu MHz		: 2199.998
cache size	: 56320 KB
physical id	: 0
siblings	: 2
core id		: 0
cpu cores	: 1
apicid		: 0
initial apicid	: 0
fpu		: yes
fpu_exception	: yes
cpuid level	: 13
wp		: yes
flags		: fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc rep_good nopl xtopology nonstop_tsc cpuid tsc_known_freq pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch invpcid_single ssbd ibrs ibpb stibp fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm rdseed adx smap xsaveopt arat md_clear arch_capabilities
bugs		: cpu_meltdown spectre_v1 spectre_v2 spec_store_bypass l1tf mds swapgs taa mmio_stale_data retbleed
bogomips	: 4399.99
clflush size	: 64
cache_alignment	: 64
addres

# I Solving the Sum of Two Digits Programming Challenge

  Gabriel

## C++

In [ ]:
%%writefile APlusB.cpp
#include <iostream>

int sum_of_two_digits(int first_digit, int second_digit) {
    return first_digit + second_digit;
}

int main() {
    int a = 0;
    int b = 0;
    std::cin >> a;
    std::cin >> b;
    std::cout << sum_of_two_digits(a, b);
    return 0;
}

Writing APlusB.cpp


In [ ]:
!g++ APlusB.cpp -o APlusB -std=c++11
!./APlusB

9 9
18

### Test input output files

In [ ]:
%%writefile input1.txt
4 8

Writing input1.txt


In [ ]:
%%writefile output1.txt
12

Writing output1.txt


In [ ]:
%%writefile input2.txt
0 9

Writing input2.txt


In [ ]:
%%writefile output2.txt
9

Writing output2.txt


In [ ]:
!./APlusB < input1.txt

12

In [ ]:
!./APlusB < input2.txt

9

### Comparing files with grep

https://askubuntu.com/questions/546796/comparing-two-text-files

In [ ]:
!grep -xvFf output1.txt output1.txt

In [ ]:
!grep -xvFf output1.txt output2.txt

9


In [ ]:
!grep -xvFf output2.txt output1.txt

12


In [ ]:
!./APlusB < input1.txt> ctest1.txt
!grep -xvFf output1.txt ctest1.txt

In [ ]:
!./APlusB < input2.txt > ctest2.txt
!grep -xvFf output2.txt ctest2.txt

## Python

In [ ]:
%%writefile sum.py
# python3

def sum_of_two_digits(first_digit, second_digit):
    return first_digit + second_digit

if __name__ == '__main__':
    a, b = map(int, input().split())
    print(sum_of_two_digits(a, b))

Writing sum.py


In [ ]:
!python sum.py

9 9
18


In [ ]:
!python sum.py  < input1.txt> ptest1.txt
!grep -xvFf output1.txt ptest1.txt

In [ ]:
!python sum.py  < input2.txt> ptest2.txt
!grep -xvFf output2.txt ptest2.txt

# II Solving The Maximum Pairwise Product Programming Challenge in C++

## C++

$$ T(n) < 6 n^2 = O(n^2)$$

In [ ]:
%%writefile max_pairwise_product_full.cpp
#include <iostream>
#include <vector>
#include <algorithm>

int MaxPairwiseProduct(const std::vector<int>& numbers) {
    int max_product = 0;
    int aux_product = 0;
    int n = numbers.size();

    for (int first = 0; first < n; ++first) {
        for (int second = 0; second < n; ++second) {
          if ( first != second ){
                aux_product = numbers[first] * numbers[second];
                if (aux_product > max_product)  max_product = aux_product;
          }
        }
    }

    return max_product;
}

int main() {
    int n;
    std::cin >> n;
    std::vector<int> numbers(n);
    for (int i = 0; i < n; ++i) {
        std::cin >> numbers[i];
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";
    return 0;
}

Writing max_pairwise_product_full.cpp


#### Time Complexity
$$\left. \begin{array}{llc}
 \text{INSTRUCTION} & \text{TYPE}& \text{# EXECUTIONS}\\
\text{first != second} &  \text{comparisons} &n^2 \\
\text{numbers[first] * numbers[second]}& \ \text{ products}  &n^2 - n \\
\text{max_product = aux_product} &  \text{assigments} & [1, \frac{n^2 - n}{2})  
\end{array}\right\}  =  O(n^2)$$

In [ ]:
!g++ max_pairwise_product_full.cpp -o max_pairwise_product_full
!./max_pairwise_product_full

^C


### Test 1: Small numbers

In [ ]:
%%writefile input0.txt
10
0 0 0 0 0 0 0 0 0 0

Writing input0.txt


In [ ]:
%%writefile output0.txt
0

Writing output0.txt


In [ ]:
!./max_pairwise_product_full< input0.txt > ctest0.txt
!grep -xvFf output0.txt ctest0.txt

In [ ]:
%%writefile input1.txt
10
1 2 3 4 5 6 7 8 9 10

Overwriting input1.txt


In [ ]:
!./max_pairwise_product < input1.txt

/bin/bash: line 1: ./max_pairwise_product: No such file or directory


In [ ]:
%%writefile output1.txt
90

Overwriting output1.txt


In [ ]:
!./max_pairwise_product_full< input1.txt > ctest1.txt
!grep -xvFf output1.txt ctest1.txt

In [ ]:
%%writefile input2.txt
10
10 9 8 7 6 5 4 3 2 1

Overwriting input2.txt


In [ ]:
%%writefile output2.txt
90

Overwriting output2.txt


In [ ]:
!./max_pairwise_product_full< input2.txt > ctest2.txt
!grep -xvFf output2.txt ctest2.txt

### Test 2: largest input numbers 200000 200000 to check possible intger overflow

In [ ]:
%%writefile input3.txt
6
1 3 200000 9 200000 23

Writing input3.txt


In [ ]:
200000*200000

40000000000

In [ ]:
%%writefile output3.txt
40000000000

Writing output3.txt


In [ ]:
!./max_pairwise_product_full < input3.txt

1345294336


In [ ]:
!./max_pairwise_product_full < input3.txt > ctest3.txt
!grep -xvFf output3.txt ctest3.txt

1345294336


In [ ]:
 %%writefile run_all_test.sh
#!/bin/bash
g++ max_pairwise_product_full.cpp -o max_pairwise_product_full
for i in {0..3}
do
    ./max_pairwise_product_full < input$i.txt > ctest$i.txt
    grep -xvFf output$i.txt ctest$i.txt
done

Writing run_all_test.sh


In [ ]:
!chmod u+x run_all_test.sh
!./run_all_test.sh

1345294336


### Changing the variables to long long

In [ ]:
%%writefile max_pairwise_product_long_full.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers) {
    long long max_product = 0LL;
	  long long aux_product = 0LL;
    int n = numbers.size();

    for (int first = 0; first < n; ++first) {
        for (int second = 0; second < n; ++second) {
          if ( first != second){
                aux_product = (long long) numbers[first] * numbers[second];
                if (aux_product > max_product)  max_product = aux_product;
          }
        }
    }


    return max_product;
}

int main() {
    int n;
    std::cin >> n;
    std::vector<int> numbers(n);
    for (int i = 0; i < n; ++i) {
        std::cin >> numbers[i];
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";
    return 0;
}


Writing max_pairwise_product_long_full.cpp


In [ ]:
!g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full

In [ ]:
!./max_pairwise_product_long_full < input3.txt

40000000000


### Runing multuple test with a bash shell

In [ ]:
 %%writefile run_all_test.sh
#!/bin/bash
g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full
for i in {1..3}
do
    ./max_pairwise_product_long_full < input$i.txt > ctest$i.txt
    grep -xvFf output$i.txt ctest$i.txt
done

Overwriting run_all_test.sh


In [ ]:
!chmod u+x run_all_test.sh
!./run_all_test.sh

### Test 3  stress tests, 6 testing cases files input4.txt/output4.txt .. input10.txt/output10.txt

---



In [ ]:
%%writefile input4.txt
7
1 3 200000 9 200000 23 200000

Writing input4.txt


In [ ]:
%%writefile output4.txt
40000000000

Writing output4.txt


In [ ]:
%%writefile input5.txt
2
1 1

Writing input5.txt


In [ ]:
%%writefile output5.txt
1

Writing output5.txt


In [ ]:
%%writefile input6.txt
8
2 2 2 2 2 2 2 2

Writing input6.txt


In [ ]:
%%writefile output6.txt
4

Writing output6.txt


In [ ]:
%%writefile input7.txt
5
1 3 200000 9 200000

Writing input7.txt


In [ ]:
%%writefile output7.txt
40000000000

Writing output7.txt


In [ ]:
%%writefile input8.txt
3
9 8 9

Writing input8.txt


In [ ]:
%%writefile output8.txt
81

Writing output8.txt


In [ ]:
%%writefile input9.txt
4
200000 200000 200000 200000

Writing input9.txt


In [ ]:
%%writefile output9.txt
40000000000

Writing output9.txt


In [ ]:
%%writefile input10.txt
0

Writing input10.txt


In [ ]:
%%writefile output10.txt
0

Writing output10.txt


In [ ]:
%%writefile run_all_test_1.sh
#!/bin/bash
g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full
for i in {4..10}
do
    ./max_pairwise_product_long_full < input$i.txt > ctest$i.txt
    grep -xvFf output$i.txt ctest$i.txt
done

Writing run_all_test_1.sh


In [ ]:
!chmod u+x run_all_test_1.sh
!./run_all_test_1.sh

### Test 4  time test  measure the execution time running with the the largest worst cases input11.txt  (200000 increasing numbers) and input12.txt (199998 increasing numbers and  200000, 200000 at the end)

In [ ]:
# input11.txt 200000 increasing numbers:  1 2 .. 200000

%%writefile WriteTest.cpp

#include <iostream>
#include <fstream>
#include <cstdlib>

using namespace std;

int main() {
  // Create and open a text file
  ofstream MyFile("input11.txt");

 MyFile << 200000 <<"\n";  // n
 for (int i=1; i <200001; i++){
     MyFile << i <<" ";
 }

  // Close the file
  MyFile.close();
}

Writing WriteTest.cpp


In [ ]:
!g++ WriteTest.cpp -o WriteTest
!./WriteTest

In [ ]:
199999 * 200000

39999800000

In [ ]:
%%writefile output11.txt
39999800000

Writing output11.txt


In [ ]:
# input12.txt 200000 decreasing numbers and 200000, 199999 ...  2 1

%%writefile WriteTest.cpp

#include <iostream>
#include <fstream>
#include <cstdlib>

using namespace std;

int main() {
  // Create and open a text file
  ofstream MyFile("input12.txt");

 MyFile << 200000 <<"\n";  // n
 for (int i=1; i <200001; i++){
     MyFile << 200001-i <<" ";
 }

  // Close the file
  MyFile.close();
}

Overwriting WriteTest.cpp


In [ ]:
!g++ WriteTest.cpp -o WriteTest
!./WriteTest

In [ ]:
%%writefile output12.txt
39999800000

Writing output12.txt


### Measure time

https://askubuntu.com/questions/53444/how-can-i-measure-the-execution-time-of-a-terminal-process

In [ ]:
!g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full
!time ./max_pairwise_product_long_full< input11.txt

^C


In [ ]:
!g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full
!time ./max_pairwise_product_long_full< input12.txt

^C


### Calculate the number or and upper bound to the number of instructions executed by the algorithm as a function of n, add in each line a comment with the number of instructions that are executed in that line.

In [ ]:
%%writefile max_pairwise_product_long_full.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers) {              //INSTRUCTIONS executed by line
    long long max_product = 0LL;                                             //1
	  long long aux_product = 0LL;                                             //1
    int n = numbers.size();                                                  //1

    for (int first = 0; first < n; ++first) {                                //2+2n
        for (int second = 0; second < n; ++second) {                         //2n + 2n**2
          if ( first != second ){                                            //n**2
                aux_product = numbers[first] * numbers[second];              //n**2 -n
                if (aux_product > max_product)                               //n**2 -n
                    max_product = aux_product;                               //[0, n**2/2 - n/2] upper bound
          }
        }
    }

                                                                             // --------------
    return max_product;                                                      // 5 + 2*n + [5*n**2, 11*n**2/2  - n/2]  upper bound
}

int main() {                                              //INSTRUCTIONS executed by line
    int n;                                                //1
    std::cin >> n;                                        //1
    std::vector<int> numbers(n);                          //n
    for (int i = 0; i < n; ++i) {                         //2 + 2n
        std::cin >> numbers[i];                           //n
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";     //1 + 1  + 5 + 2*n + [5*n**2, 11*n**2/2  - n/2]

    return 0;                                             // --------------
}                                                         //  11 + 6*n + [5*n**2, 11*n**2/2  - n/2]  upper bound


Overwriting max_pairwise_product_long_full.cpp


### Count executed intructions in code

In [ ]:
%%writefile max_pairwise_product_long_full_execints.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers, long long &execints) {      //INSTRUCTIONS executed by line

    long long max_product = 0LL;                                                          //1
    execints++;

	  long long aux_product = 0LL;                                                          //1
    execints++;

    int n = numbers.size();                                                               //1
    execints++;

    execints++;
    for (int first = 0; first < n; ++first) {                                             //2+2n
        execints += 2;

        execints++;
        for (int second = 0; second < n; ++second) {                                       //2n + 2n**2
          execints += 2;

          execints++;
          if ( first != second ){                                                         //n**2
                aux_product = numbers[first] * numbers[second];                           //n**2 -n
                execints++;

                execints++;
                if (aux_product > max_product){                                           //n**2 -n
                    max_product = aux_product;                                            //[0,n**2/2 - n/2)  upper bound
                    execints++;

                }
        }

      }
      execints++;
    }
    execints++;
                                                                                        // --------------
    return max_product;                                                                 // 5 + 2*n + [5*n**2, 11*n**2/2  - n/2] upper bound
}

int main() {                                                 //INSTRUCTIONS executed by line

    long long execints =0LL;

    int n;                                                   //1
    execints++;

    std::cin >> n;                                           //1
    execints++;

    std::vector<int> numbers(n);                             //n
    execints += n;

    execints++;
    for (int i = 0; i < n; ++i) {                            //2 + 2n
        execints += 2;

        std::cin >> numbers[i];
        execints++;                                          //n

    }


    //std::cout << MaxPairwiseProduct(numbers) << "\n";      //1 + 1  + 5 + 2*n + [5*n**2, 11*n**2/2  - n/2] upper bound

    execints++;
    int M = MaxPairwiseProduct(numbers,execints);

    execints++;
    std::cout <<  execints << "\n";
    return 0;                                                // --------------
}                                                            // 11 + 6*n + [5*n**2, 11*n**2/2  - n/2]

Writing max_pairwise_product_long_full_execints.cpp


The number of instructions executed by the algooirithm with a list of size $n$ is  

$$T (n) = 11 + 6n + \left[5n^2, 11\frac{n^2}{2}  - \frac{n}{2}\right)  >  6n^2 = O(n^2)$$

It depends on the size of the lists and  the order of the list.

For the input0.txt
```
10
0 0 0 0 0 0 0 0 0 0
```
$$T = 11 + 6(10) + 5(10)^2   = 561 $$

$$T(10) \in [571, 621)  $$


In [ ]:
!g++ max_pairwise_product_long_full_execints.cpp -o max_pairwise_product_long_full_execints
!./max_pairwise_product_long_full_execints < input0.txt

570


For the input1.txt
```
10
1 2 3 4 5 6 7 8 9 10
```
$$T = 11 + 6(10) + 5(10)^2 +  9 + 5 + 4 + 3 + 2 + 2 + 2 + 2 + 1 + 0  = 601 $$

$$T(10) \in [572, 622)  $$


In [ ]:
#The number of times that  max_product = aux_product; in the worst case is less than (n**2 -n)/2 -  upper bound

#     1  2  3  4  5  6  7  8  9  10
#  1  x  2  3  4  5  6  7  8  9  10         9
#  2  2  x  6  8  10 12 14 16 18 20         5 14
#  3  3  6  x  12 15 18 21 24 27 30         4 18
#  4  4  8  12 x  20 24 28 32 36 40         3 21
#  5  5  10 15 20 x  30 35 40 45 50         2 23
#  6  6  12 18 24 30 x  42 48 54 60         2 25
#  7  7  14 21 28 35 42 x  56 63 70         2 27
#  8  8  18 27 36 45 54 63 72 x  90         1 30
#  10 10 20 30 40 50 60 70 60 90 x          0

In [ ]:
11 + 6*(10) + 5*(10)*(10) +  9 + 5 + 4 + 3 + 2 + 2 + 2 + 2 + 1 + 0

601

In [ ]:
n = 10
print( "[", 11 + 6*n + 5*n**2,  ",", 11 + 6*n  + 5.5*n**2, ")")

[ 571 , 621.0 )


In [ ]:
!g++ max_pairwise_product_long_full_execints.cpp -o max_pairwise_product_long_full_execints
!./max_pairwise_product_long_full_execints < input1.txt

600



For the input2.txt
```
10
10 9 8 7 6 5 4 3 2 1
```
$$T = 11 + 6*(10) + 5*(10)**2 +  1  = 571 $$

In [ ]:
11 + 6*(10) + 5*(10)**2 +  1

572

In [ ]:
!g++ max_pairwise_product_long_full_execints.cpp -o max_pairwise_product_long_full_execints
!./max_pairwise_product_long_full_execints < input2.txt

571


Calculate estimated time by instruction and number of instructions by second

For the input11.txt
```
200000
1 2 ..  199999 2000
```
$$T = 11 + 6*(200000) + 5.5*(200000)**2  = 200001200012  $$

In [ ]:
!g++ max_pairwise_product_long_full_execints.cpp -o max_pairwise_product_long_full_execints
!./max_pairwise_product_long_full_execints < input11.txt

^C


In [ ]:
# Estimated time by instruction

295/200003177092


1.4749765693187072e-09

In [ ]:
# Estimated number of instructions by second

200003177092/295

677976871.4983051

### Estimate and verify the maximun n that can be solved in 1 second

$$(5.5n^2 + 5.5n + 11) \times 1.47*10^{-9} s = 1 s$$

$$ n \approx 11000$$


In [ ]:
# Verify input13.txt
%%writefile WriteTest.cpp

#include <iostream>
#include <fstream>
#include <cstdlib>

using namespace std;

int main() {
  // Create and open a text file
  ofstream MyFile("input13.txt");

 MyFile << 11000 <<"\n";  // n
 for (int i=1; i <11001; i++){
     MyFile << i <<" ";
 }

  // Close the file
  MyFile.close();
}

Overwriting WriteTest.cpp


In [ ]:
!g++ WriteTest.cpp -o WriteTest
!./WriteTest

In [ ]:
10999 * 11000

120989000

In [ ]:
!g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full
!time ./max_pairwise_product_long_full< input13.txt

120989000

real	0m0.824s
user	0m0.814s
sys	0m0.003s


## Python

Breiner


In [ ]:
%%writefile max_pairwise_product_long_full.py
def MaxPairwiseProduct(numbers):
  max_product = 0
  aux_product = 0
  n = len(numbers)

  for first in range (n):
    for second in range(n):
      if(first != second):
        aux_product = numbers[first]*numbers[second]
        if aux_product > max_product:
          max_product = aux_product
  return max_product

n = int(input())
numbers = [int(i) for i in input().split()]
print(MaxPairwiseProduct(numbers))


Writing max_pairwise_product_long_full.py


In [ ]:
# Verify
!time python3 max_pairwise_product_long_full.py < input13.txt

Traceback (most recent call last):
  File "/content/max_pairwise_product_long_full.py", line 16, in <module>
    print(MaxPairwiseProduct(numbers))
  File "/content/max_pairwise_product_long_full.py", line 10, in MaxPairwiseProduct
    if aux_product > max_product:
KeyboardInterrupt
^C


# For the follwing solutions rquired  the same tests and analysis performed for the first solution

## Solutions in C++ and Python that uses
$$\frac{n^2 - n}{2} = O(n^2)\ \ \text{products}$$

In [ ]:
%%writefile max_pairwise_product.cpp
#include <iostream>
#include <vector>
#include <algorithm>

int MaxPairwiseProduct(const std::vector<int>& numbers) {
    int max_product = 0;
    int n = numbers.size();

    for (int first = 0; first < n; ++first) {
        for (int second = first + 1; second < n; ++second) {
            max_product = std::max(max_product,
                numbers[first] * numbers[second]);
        }
    }

    return max_product;
}

int main() {
    int n;
    std::cin >> n;
    std::vector<int> numbers(n);
    for (int i = 0; i < n; ++i) {
        std::cin >> numbers[i];
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";
    return 0;
}

Writing max_pairwise_product.cpp


In [ ]:
%%writefile max_pairwise_product_long.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers) {              //INSTRUCTIONS executed by line
    long long max_product = 0LL;                                             //1
	  long long aux_product = 0LL;                                             //1
    int n = numbers.size();                                                  //1

    for (int first = 0; first < n; ++first) {                                //2+2n
        for (int second = first + 1; second < n; ++second) {                 //2n + (n-1)n
			       aux_product =  (long long) numbers[first] * numbers[second];    //(n-1)n/2
            if (max_product < aux_product)  {                                //(n-1)n/2
				        max_product = aux_product;                                   //[1,(n-1)n/2)  upper bound
             }
        }
    }
                                                                             // --------------
    return max_product;                                                      // 5 + 4n + [2(n-1)n+1, 2.5 (n-1)n) upper bound
}

int main() {                                              //INSTRUCTIONS executed by line
    int n;                                                //1
    std::cin >> n;                                        //1
    std::vector<int> numbers(n);                          //n
    for (int i = 0; i < n; ++i) {                         //2 + 2n
        std::cin >> numbers[i];                           //n
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";     //1 + 1  + 5 + 4n + [2(n-1)n+1, 2.5 (n-1)n) upper bound
    return 0;                                             // --------------
}                                                         // 11 + 7n + [2(n-1)n+1, 2.5 (n-1)n) upper bound


Writing max_pairwise_product_long.cpp


In [ ]:
!g++ max_pairwise_product_long.cpp -o max_pairwise_product_long
!time ./max_pairwise_product_long < input13.txt

120989000

real	0m0.383s
user	0m0.379s
sys	0m0.001s


### Estimate and verify the maximun n that can be solved in 1 second



In [ ]:
!g++ max_pairwise_product.cpp -o max_pairwise_product
!time ./max_pairwise_product < input13.txt

120989000

real	0m0.571s
user	0m0.555s
sys	0m0.005s


# Sort solution in C++ and Python

$$ O(n \lg n)$$

**Which sorting algorithm is used in GCC?**
https://stackoverflow.com/questions/7221257/which-sorting-algorithm-is-used-in-gcc

* ...  the algorithm is called introsort, a combination between Insertionsort, Quicksort & Heapsort. ...

**Introsort**
https://en.wikipedia.org/wiki/Introsort

Introsort or introspective sort is a hybrid sorting algorithm that provides both fast average performance and (asymptotically) optimal worst-case performance. It begins with quicksort, it switches to heapsort when the recursion depth exceeds a level based on (the logarithm of) the number of elements being sorted and it switches to insertion sort when the number of elements is below some threshold

*   Class	Sorting algorithm
*   Data structure	Array
*   Worst-case performance	$O(n \lg n)$
*   Average performance	$O(n \lg n)$

**Tinsort**
https://en.wikipedia.org/wiki/Timsort

Timsort is a hybrid, stable sorting algorithm, derived from merge sort and insertion sort, designed to perform well on many kinds of real-world data. It was implemented by Tim Peters in 2002 for use in the Python programming language.

*   Class	Sorting algorithm
*   Data structure	Array
*   Worst-case performance	$O(n \lg n)$
*   Best-case performance $O(n)$
*   Average performance	$O(n \lg n)$
*   Worst-case space complexity $O(n)$

In [ ]:

%%writefile max_pairwise_product_with_sort.py
#Python3
if __name__ == "__main__":
    n = int(input())
    lista = list(map(int, input().split()))
    lista.sort()
    print(lista[n - 1] * lista[n - 2])

Writing max_pairwise_product_with_sort.py



Overwriting max_pairwise_product_with_sort.py

In [ ]:

%%writefile max_pairwise_product_with_sort_C.cpp
#include
#include
#include

int main() {
    int n;
    std::cin >> n;
    std::vector numbers(n);
    for (int i = 0; i < n; ++i) {
        std::cin >> numbers[i];
    }

    std::sort(numbers.begin(), numbers.end());
    std::cout << (long long) numbers[n - 1] * numbers[n - 2] << '\n';

    return 0;


}


Writing max_pairwise_product_with_sort_C.cpp


In [ ]:

!g++ max_pairwise_product_with_sort_C.cpp -o max_pairwise_product_with_sort_C

max_pairwise_product_with_sort_C.cpp:1:9: error: #include expects "FILENAME" or <FILENAME>
    1 | #include
      |         ^
max_pairwise_product_with_sort_C.cpp:2:9: error: #include expects "FILENAME" or <FILENAME>
    2 | #include
      |         ^
max_pairwise_product_with_sort_C.cpp:3:9: error: #include expects "FILENAME" or <FILENAME>
    3 | #include
      |         ^
max_pairwise_product_with_sort_C.cpp: In function ‘int main()’:
max_pairwise_product_with_sort_C.cpp:7:10: error: ‘cin’ is not a member of ‘std’
    7 |     std::cin >> n;
      |          ^~~
max_pairwise_product_with_sort_C.cpp:1:1: note: ‘std::cin’ is defined in header ‘<iostream>’; did you forget to ‘#include <iostream>’?
  +++ |+#include <iostream>
    1 | #include
max_pairwise_product_with_sort_C.cpp:8:10: error: ‘vector’ is not a member of ‘std’
    8 |     std::vector numbers(n);
      |          ^~~~~~
max_pairwise_product_with_sort_C.cpp:1:1: note: ‘std::vector’ is defined in header ‘<vector>’; did you fo

Test with inputs 1 and 2



In [ ]:


!./max_pairwise_product_with_sort_C < input1.txt
!./max_pairwise_product_with_sort_C < input2.txt


/bin/bash: line 1: ./max_pairwise_product_with_sort_C: No such file or directory
/bin/bash: line 1: ./max_pairwise_product_with_sort_C: No such file or directory


Test and time it with input 3

In [ ]:
!time ./max_pairwise_product_with_sort_C < input3.txt

/bin/bash: line 1: ./max_pairwise_product_with_sort_C: No such file or directory

real	0m0.001s
user	0m0.000s
sys	0m0.001s


In [ ]:
%%time
!python max_pairwise_product_with_sort.py < input3.txt

40000000000
CPU times: user 5.6 ms, sys: 933 µs, total: 6.53 ms
Wall time: 107 ms


20 stress tests

In [ ]:
%%writefile stress_tests.py
#Python3
import random

if __name__ == "__main__":
    for i in range(20):
        n = random.randint(2, 99)
        lista = [random.randint(17, 25631) for i in range(n)]
        lista.sort()
        print(f"n: {n} Product: {lista[n - 1] * lista[n - 2]}")



Writing stress_tests.py


In [ ]:
%%writefile stress_tests_C.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <random>

int main() {
    std::random_device rd;
    std::mt19937 gen(rd());
    std::uniform_int_distribution<int> n_dist(2, 99);
    std::uniform_int_distribution<int> num_dist(0, 25000);

    for (int i = 0; i < 20; ++i ){
          int n = n_dist(gen);                              //1
          std::vector<int> numbers(n);                          //n
          for (int i = 0; i < n; ++i) {                         //1 + 2n
              numbers[i]= num_dist(gen);                           //n
          }

          std::sort(numbers.begin(), numbers.end());                          //nlogn
          std::cout << "n= "<< n << " Product= " << (long long) numbers[n - 1] * numbers[n - 2] << '\n';   //2
    }
    return 0;
}

Writing stress_tests_C.cpp


In [ ]:
!g++ stress_tests_C.cpp -o stress_tests_C
!./stress_tests_C

n= 63 Product= 589831480
n= 27 Product= 546510899
n= 25 Product= 572670301
n= 71 Product= 619551272
n= 87 Product= 615179784
n= 88 Product= 573367788
n= 61 Product= 589019600
n= 12 Product= 501835044
n= 45 Product= 580006258
n= 34 Product= 611078231
n= 59 Product= 606371025
n= 99 Product= 616577832
n= 61 Product= 578143256
n= 18 Product= 427741661
n= 17 Product= 377867882
n= 16 Product= 473052327
n= 4 Product= 384746796
n= 49 Product= 606747570
n= 29 Product= 585016192
n= 9 Product= 457106654


In [ ]:
!python stress_tests.py

n: 94 Product: 649000108
n: 86 Product: 646171071
n: 34 Product: 630156990
n: 46 Product: 595831200
n: 71 Product: 613169688
n: 98 Product: 641988750
n: 93 Product: 608585256
n: 55 Product: 637777560
n: 49 Product: 608122990
n: 43 Product: 638293700
n: 84 Product: 613748160
n: 51 Product: 606234564
n: 59 Product: 603363540
n: 47 Product: 554812960
n: 43 Product: 620271720
n: 69 Product: 616577680
n: 38 Product: 575373504
n: 15 Product: 447983032
n: 45 Product: 634635100
n: 24 Product: 525926016


Count executed instructions in code and verify with inputs 1,2 and 3


In [ ]:
%%writefile max_pairwise_product_with_sort_and_instructions.py
#Python3
import math
if __name__ == "__main__":
    inst=0
    n = int(input())
    inst+=1
    lista = list(map(int, input().split()))
    inst+= 4*n
    lista.sort()
    inst+=  n * math.log2(n)
    print(lista[n - 1] * lista[n - 2])
    inst+=1
    print(f"Instrucciones ejecutadas: {int(inst)}\n")

Writing max_pairwise_product_with_sort_and_instructions.py


In [ ]:
%%writefile max_pairwise_product_with_sort_C_and_instructions.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <cmath>

int main() {
    int inst;
    int n;                                                //1
    inst++;
    std::cin >> n;                                        //1
    inst++;
    std::vector<int> numbers(n);                          //n
    inst+=n;
    inst+=1;                                              // add the declaration of i=0
    for (int i = 0; i < n; ++i) {                         //1 + 2n
        inst+=2;
        std::cin >> numbers[i];                           //n
        inst++;
    }

    std::sort(numbers.begin(), numbers.end());                          //nlogn
    inst+=n * log2(n);
    std::cout << (long long) numbers[n - 1] * numbers[n - 2] << '\n';   //2
    inst+=2;
    std::cout << "Instrucciones Ejecutadas: " << inst << '\n';
    return 0;


}

Writing max_pairwise_product_with_sort_C_and_instructions.cpp


In [ ]:
!g++ max_pairwise_product_with_sort_C_and_instructions.cpp -o max_pairwise_product_with_sort_C_and_instructions

In [ ]:
!./max_pairwise_product_with_sort_C_and_instructions < input1.txt
!./max_pairwise_product_with_sort_C_and_instructions < input2.txt
!./max_pairwise_product_with_sort_C_and_instructions < input3.txt

90
Instrucciones Ejecutadas: 78
90
Instrucciones Ejecutadas: 78
40000000000
Instrucciones Ejecutadas: 44


In [ ]:
!python max_pairwise_product_with_sort_and_instructions.py < input1.txt
!python max_pairwise_product_with_sort_and_instructions.py < input2.txt
!python max_pairwise_product_with_sort_and_instructions.py < input3.txt

90
Instrucciones ejecutadas: 75

90
Instrucciones ejecutadas: 75

40000000000
Instrucciones ejecutadas: 41



Calculate estimated time by instruction and number of instructions by second

In [ ]:
%%writefile max_pairwise_product_with_sort_and_instructions_with_time_and_IbS.py
#Python3
import math
import time
if __name__ == "__main__":
    inicio = time.time()
    inst=0
    n = int(input())
    inst+=1
    lista = list(map(int, input().split()))
    inst+= 4*n
    lista.sort()
    inst+=  n * math.log2(n)
    print(lista[n - 1] * lista[n - 2])
    inst+=1
    print(f"Instrucciones ejecutadas: {int(inst)}")
    fin = time.time()
    tiempo_total = fin - inicio
    instructionTime = tiempo_total/inst
    instructionBySecond= inst/tiempo_total

    print(f"Tiempo total de ejecución: {tiempo_total*1000} milisegundos")
    print(f"Tiempo estimado de cada instruccion: {instructionTime*1000} milisegundos")
    print(f"Instrucciones por segundo: {int(instructionBySecond)} ")

Writing max_pairwise_product_with_sort_and_instructions_with_time_and_IbS.py


### Estimate and verify the maximun n that can be solved in 1 second

In [ ]:
!python max_pairwise_product_with_sort_and_instructions_with_time_and_IbS.py < input13.txt

120989000
Instrucciones ejecutadas: 191679
Tiempo total de ejecución: 2.3975372314453125 milisegundos
Tiempo estimado de cada instruccion: 1.2508060568551837e-05 milisegundos
Instrucciones por segundo: 79948445 


## Efficient solution with largest and second-largest in an list, array or vector in C++ and Python

 $$T(n) < 10 n  = O(n)$$





Add in each line a coment with the number of intstrcution that are executed in that line

In [ ]:
%%writefile max_pairwise_product_long_efficient.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers) {
    long long max_product = 0LL;                                             //1
	  long long n_major = 0LL;
    long long n_major_2 = 0LL;                                             //1
    int n = numbers.size();                                                  //1

    for(int i = 0; i < n; ++i) {                   //1 + 2n
      if (numbers[i] > n_major){
        n_major_2 = n_major ;                //n
        n_major = numbers[i];
      }
      else if (numbers[i] > n_major_2){                   //n
        n_major_2 = numbers[i];                    //[0, n]

      }
    }
    max_product=n_major*n_major_2;                                                                          // --------------
    return max_product;
}

int main() {
    int n;                                                //1
    std::cin >> n;                                        //1
    std::vector<int> numbers(n);                          //n
    for (int i = 0; i < n; ++i) {                         //1 + 2n
        std::cin >> numbers[i];                           //n
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";
    return 0;
}

Writing max_pairwise_product_long_efficient.cpp


In [ ]:
!g++ max_pairwise_product_long_efficient.cpp -o max_pairwise_product_long_efficient

In [ ]:
%%writefile max_pairwise_product_efficient.py
#Python3
def max_pairwise_product(lista):
    n_major = 0
    n_major_2 = 0
    max_product = 0
    for i in lista:
        if i > n_major:
            n_major_2=n_major
            n_major = i
        elif i > n_major_2:
            n_major_2=i
    max_product = n_major_2 * n_major

    return max_product

if __name__ == "__main__":
    n = int(input())
    lista = list(map(int, input().split()))
    print(max_pairwise_product(lista))

Writing max_pairwise_product_efficient.py


Test with inputs 1 and 2



In [ ]:
!./max_pairwise_product_long_efficient < input1.txt
!./max_pairwise_product_long_efficient < input2.txt


90
90


In [ ]:
!python max_pairwise_product_efficient.py < input1.txt
!python max_pairwise_product_efficient.py < input2.txt

90
90


Test and time it with input 3

In [ ]:
%%time
!python max_pairwise_product_efficient.py < input3.txt

40000000000
CPU times: user 4.82 ms, sys: 1.13 ms, total: 5.95 ms
Wall time: 104 ms


In [ ]:
!time ./max_pairwise_product_long_efficient < input3.txt

40000000000

real	0m0.003s
user	0m0.000s
sys	0m0.003s


20 stress tests

In [ ]:
%%writefile max_pairwise_product_long_efficient_stress_test.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <random>

long long MaxPairwiseProduct(const std::vector<int>& numbers) {
    long long max_product = 0LL;                                             //1
	  long long n_major = 0LL;
    long long n_major_2 = 0LL;                                             //1
    int n = numbers.size();                                                  //1

    for(int i = 0; i < n; ++i) {                   //1 + 2n
      if (numbers[i] > n_major){
        n_major_2 = n_major ;                //n
        n_major = numbers[i];
      }
      else if (numbers[i] > n_major_2){                   //n
        n_major_2 = numbers[i];                    //[0, n]

      }
    }
    max_product=n_major*n_major_2;                                                                          // --------------
    return max_product;
}

int main() {
    std::random_device rd;
    std::mt19937 gen(rd());
    std::uniform_int_distribution<int> n_dist(2, 99);
    std::uniform_int_distribution<int> num_dist(0, 25000);
    int n;                                                //1
    std::vector<int> numbers(n);                          //n
    for (int i = 0; i < 20; ++i ){
      int n = n_dist(gen);                              //1
      std::vector<int> numbers(n);                          //n
      for (int i = 0; i < n; ++i) {                         //1 + 2n
        numbers[i]= num_dist(gen);                           //n
      }
      std::cout << "n= "<< n << " Product= " << MaxPairwiseProduct(numbers) << "\n";
    }
    return 0;
}

Writing max_pairwise_product_long_efficient_stress_test.cpp


In [ ]:
!g++ max_pairwise_product_long_efficient_stress_test.cpp -o max_pairwise_product_long_efficient_stress_test
!./max_pairwise_product_long_efficient_stress_test

n= 48 Product= 586881456
n= 47 Product= 602784784
n= 57 Product= 584603064
n= 44 Product= 589318847
n= 96 Product= 614171604
n= 99 Product= 557194785
n= 77 Product= 609212000
n= 4 Product= 325104422
n= 15 Product= 507461583
n= 15 Product= 596612648
n= 94 Product= 618476028
n= 17 Product= 514403454
n= 81 Product= 608738130
n= 46 Product= 557110554
n= 2 Product= 272983822
n= 84 Product= 610680076
n= 50 Product= 520618846
n= 21 Product= 540663500
n= 49 Product= 612423504
n= 37 Product= 549594507


In [ ]:
%%writefile max_pairwise_product_efficient_stress_tests.py
#Python3
import random
def max_pairwise_product(lista):
    n_major = 0
    n_major_2 = 0
    max_product = 0
    for i in lista:
        if i > n_major:
            n_major_2=n_major
            n_major = i
        elif i > n_major_2:
            n_major_2=i
    max_product = n_major_2 * n_major

    return max_product

if __name__ == "__main__":
    for i in range(20):
      n = random.randint(2, 99)
      lista = [random.randint(17, 25631) for i in range(n)]
      print(f"n=  {n}   Product= {max_pairwise_product(lista)}")

Writing max_pairwise_product_efficient_stress_tests.py


In [ ]:
%%writefile max_pairwise_product_efficient_stress_tests.py
#Python3
import random
def max_pairwise_product(lista):
    n_major = 0
    n_major_2 = 0
    max_product = 0
    for i in lista:
        if i > n_major:
            n_major_2=n_major
            n_major = i
        elif i > n_major_2:
            n_major_2=i
    max_product = n_major_2 * n_major

    return max_product

if __name__ == "__main__":
    for i in range(20):
      n = random.randint(2, 99)
      lista = [random.randint(17, 25631) for i in range(n)]
      print(f"n=  {n}   Product= {max_pairwise_product(lista)}")

Overwriting max_pairwise_product_efficient_stress_tests.py


Count executed instructions in code and verify with inputs 1,2 and 3


In [ ]:
%%writefile max_pairwise_product_long_efficient_and_instructions.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long inst= 0LL;

long long MaxPairwiseProduct(const std::vector<int>& numbers) {
    long long max_product = 0LL;                                             //1
    inst++;
	  long long n_major = 0LL;
    inst++;
    long long n_major_2 = 0LL;                                             //1
    inst++;
    int n = numbers.size();                                                  //1
    inst++;
    inst++;
    for(int i = 0; i < n; ++i) {                   //1 + 2n
      inst+=2;
      inst++;
      if (numbers[i] > n_major){
        inst++;
        n_major_2 = n_major ;                //n
        inst++;
        n_major = numbers[i];
        inst++;
      }
      else if (numbers[i] > n_major_2){
        inst++;
        n_major_2 = numbers[i];                    //[0, n]
        inst++;
        inst++;
      }
    }
    max_product=n_major*n_major_2;
    inst++;                                                                         // --------------
    return max_product;
}

int main() {
    int n;                                                //1
    inst++;
    std::cin >> n;                                        //1
    inst++;
    std::vector<int> numbers(n);                          //n
    inst++;
    inst++;
    for (int i = 0; i < n; ++i) {                         //1 + 2n
        std::cin >> numbers[i];                           //n
    }
    inst+=n;

    std::cout << MaxPairwiseProduct(numbers) << "\n";
    inst++;
    std::cout <<"Numero de instrucciones: "<< inst << "\n";
    return 0;
}

Writing max_pairwise_product_long_efficient_and_instructions.cpp


In [ ]:
!g++ max_pairwise_product_long_efficient_and_instructions.cpp -o max_pairwise_product_long_efficient_and_instructions


In [ ]:
!./max_pairwise_product_long_efficient_and_instructions < input1.txt
!./max_pairwise_product_long_efficient_and_instructions < input2.txt
!./max_pairwise_product_long_efficient_and_instructions < input3.txt

90
Numero de instrucciones: 81
90
Numero de instrucciones: 57
40000000000
Numero de instrucciones: 50


In [ ]:
%%writefile max_pairwise_product_efficient_and_instructions.py
#Python3

inst=0
def max_pairwise_product(lista):
    global inst
    n_major = 0
    inst+=1
    n_major_2 = 0
    inst+=1
    max_product = 0
    inst+=1
    for i in lista:
        inst+=1
        if i > n_major:
            inst+=1
            n_major_2=n_major
            inst+=1
            n_major = i
            inst+=1
        elif i > n_major_2:
            inst+=1
            n_major_2=i
            inst+=1
            inst+=1
    max_product = n_major_2 * n_major
    inst+=1

    return max_product

if __name__ == "__main__":
    n = int(input())
    inst+=1
    lista = list(map(int, input().split()))
    inst+=4*n
    print(max_pairwise_product(lista))
    inst+=1
    print(f"Numero Total de Instrucciones:  {inst}")

Writing max_pairwise_product_efficient_and_instructions.py


In [ ]:
!python max_pairwise_product_efficient_and_instructions.py < input1.txt
!python max_pairwise_product_efficient_and_instructions.py < input2.txt
!python max_pairwise_product_efficient_and_instructions.py < input3.txt

90
Numero Total de Instrucciones:  86
90
Numero Total de Instrucciones:  62
40000000000
Numero Total de Instrucciones:  51


Calculate estimated time by instruction and number of instructions by second

In [ ]:
%%writefile max_pairwise_product_long_efficient_and_instructions_by_second.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <chrono>

long long inst= 0LL;

long long MaxPairwiseProduct(const std::vector<int>& numbers) {
    long long max_product = 0LL;                                             //1
    inst++;
	  long long n_major = 0LL;
    inst++;
    long long n_major_2 = 0LL;                                             //1
    inst++;
    int n = numbers.size();                                                  //1
    inst++;
    inst++;
    for(int i = 0; i < n; ++i) {                   //1 + 2n
      inst+=2;
      inst++;
      if (numbers[i] > n_major){
        inst++;
        n_major_2 = n_major ;                //n
        inst++;
        n_major = numbers[i];
        inst++;
      }
      else if (numbers[i] > n_major_2){
        inst++;
        n_major_2 = numbers[i];                    //[0, n]
        inst++;
        inst++;
      }
    }
    max_product=n_major*n_major_2;
    inst++;                                                                         // --------------
    return max_product;
}

int main() {
    auto start_time = std::chrono::high_resolution_clock::now();
    int n;                                                //1
    inst++;
    std::cin >> n;                                        //1
    inst++;
    std::vector<int> numbers(n);                          //n
    inst++;
    inst++;
    for (int i = 0; i < n; ++i) {                         //1 + 2n
        std::cin >> numbers[i];                           //n
    }
    inst+=n;

    std::cout << MaxPairwiseProduct(numbers) << "\n";
    inst++;
    auto end_time = std::chrono::high_resolution_clock::now();
    auto duration = std::chrono::duration_cast<std::chrono::milliseconds>(end_time - start_time);
    float duration_sec,duration_f;
    duration_f=static_cast<float>(duration.count());
    duration_sec = duration_f/1000;
    std::cout <<"Numero de instrucciones: "<< inst << "\n";
    std::cout <<"Duracion Estiamda: "<< duration_sec << " segundos" <<"\n";
    std::cout <<"Duracion Estiamda por instruccion: "<< duration_sec/inst << " segundos" <<"\n";
    std::cout <<"Numero de instrucciones por segundo: "<< 1/(duration_sec/inst) << " Instrucciones" <<"\n";
    return 0;
}

Writing max_pairwise_product_long_efficient_and_instructions_by_second.cpp


In [ ]:
!g++ max_pairwise_product_long_efficient_and_instructions_by_second.cpp -o max_pairwise_product_long_efficient_and_instructions_by_second

In [ ]:
!./max_pairwise_product_long_efficient_and_instructions_by_second < input13.txt

120989000
Numero de instrucciones: 77011
Duracion Estiamda: 0.004 segundos
Duracion Estiamda por instruccion: 5.19406e-08 segundos
Numero de instrucciones por segundo: 1.92527e+07 Instrucciones


### Estimate and verify the maximun n that can be solved in 1 second

In [ ]:
%%writefile max_pairwise_product_efficient_and_instructions_by_second.py
#Python3
import time

inst=0
def max_pairwise_product(lista):
    global inst
    n_major = 0
    inst+=1
    n_major_2 = 0
    inst+=1
    max_product = 0
    inst+=1
    for i in lista:
        inst+=1
        if i > n_major:
            inst+=1
            n_major_2=n_major
            inst+=1
            n_major = i
            inst+=1
        elif i > n_major_2:
            inst+=1
            n_major_2=i
            inst+=1
            inst+=1
    max_product = n_major_2 * n_major
    inst+=1

    return max_product

if __name__ == "__main__":
    inicio=time.time()
    n = int(input())
    inst+=1
    lista = list(map(int, input().split()))
    inst+=4*n
    print(max_pairwise_product(lista))
    inst+=1
    fin = time.time()
    tiempo_total = fin - inicio
    instructionTime = tiempo_total/inst
    instructionBySecond= 1/instructionTime
    print(f"Numero Total de Instrucciones:  {inst}")
    print(f"Tiempo total de ejecución: {tiempo_total} segundos")
    print(f"Tiempo estimado de cada instruccion: {instructionTime} segundos")
    print(f"Instrucciones por segundo: {int(instructionBySecond)} ")


Writing max_pairwise_product_efficient_and_instructions_by_second.py


In [ ]:
!g++ max_pairwise_product_long_efficient_and_instructions_by_second.cpp -o max_pairwise_product_long_efficient_and_instructions_by_second

In [ ]:
!./max_pairwise_product_long_efficient_and_instructions_by_second < input13.txt

120989000
Numero de instrucciones: 77011
Duracion Estiamda: 0.005 segundos
Duracion Estiamda por instruccion: 6.49258e-08 segundos
Numero de instrucciones por segundo: 1.54022e+07 Instrucciones


## Super efficient solution

Whitout storing the list in memory, only stpring the largest and second largest and whitout calling a function, add in each line a coment with the number of instructions that are executed in that line

in C++ and Python


 $$T(n) < 13 n  = O(n)$$


In [ ]:
%%writefile max_pairwise_sort_super_efficient.py
n = int(input())                            # 1
max_number = float("-inf")                  # 1
second_max_number = float("-inf")           # 1

for num in input().split():                 # 3 + 2n

    num = int(num)                          # n

    if num > max_number:                    # n
        second_max_number = max_number      # [0, n] in the worst case it would be n times
        max_number = num                    # [0, n] in the worst case it would be n times
    elif num > second_max_number:           # n
        second_max_number = num             # [0, n] in the worst case it would be n times

print(f"n: {n} Product: {max_number * second_max_number}")

Writing max_pairwise_sort_super_efficient.py


Test with inputs 1 and 2




In [ ]:
!python max_pairwise_sort_super_efficient.py < input1.txt
!python max_pairwise_sort_super_efficient.py < input2.txt

n: 10 Product: 90
n: 10 Product: 90


Test and time it with input 3

In [ ]:
%%time
!python  max_pairwise_sort_super_efficient.py < input3.txt

n: 6 Product: 40000000000
CPU times: user 6.41 ms, sys: 912 µs, total: 7.33 ms
Wall time: 107 ms


20 stress tests

In [ ]:
%%writefile max_pairwise_sort_super_efficient_stress_test.py
n = int(input())                            # 1
max_number = float("-inf")                  # 1
second_max_number = float("-inf")           # 1

for num in input().split():                 # 3 + 2n

    num = int(num)                          # n

    if num > max_number:                    # n
        second_max_number = max_number      # [0, n] in the worst case it would be n times
        max_number = num                    # [0, n] in the worst case it would be n times
    elif num > second_max_number:           # n
        second_max_number = num             # [0, n] in the worst case it would be n times

print(f"n: {n} Product: {max_number * second_max_number}")

def stress_test(num_tests, max_n, max_value):
    for i in range(num_tests):
        n = random.randint(2, max_n)
        nums = [random.randint(1, max_value) for _ in range(n)]
        start_time = time.time()
        product = max_pairwise_product(nums)
        end_time = time.time()
        print(f"Test {i+1}: n = {n}, Product = {product}, Time = {end_time - start_time:.6f} seconds")

if __name__ == "__main__":
    num_tests = 20
    max_n = 10**5
    max_value = 10**9
    stress_test(num_tests, max_n, max_value)

Writing max_pairwise_sort_super_efficient_stress_test.py


Count executed instructions in code and verify with inputs 1,2 and 3

In [ ]:
%%writefile max_pairwise_sort_super_efficient_instructions.py
instruction_count = 0

n = int(input())                            # 1
instruction_count += 1
max_number = float("-inf")                  # 1
instruction_count += 1
second_max_number = float("-inf")           # 1
instruction_count += 1

# Iterating over range of n
for _ in range(n):                          # 1 + n
    instruction_count += 1
    num = int(input())                      # n
    instruction_count += 1

    if num > max_number:                    # n
        instruction_count += 1
        second_max_number = max_number      # [0, n] in the worst case it would be n times
        instruction_count += 1
        max_number = num                    # [0, n] in the worst case it would be n times
        instruction_count += 1
    elif num > second_max_number:           # n
        instruction_count += 1
        second_max_number = num             # [0, n] in the worst case it would be n times
        instruction_count += 1

print(f"n: {n} Product: {max_number * second_max_number}")
instruction_count += 1

print("Total instructions executed:", instruction_count)

Writing max_pairwise_sort_super_efficient_instructions.py


Calculate estimated time by instruction and number of instructions by second

In [ ]:
%%writefile max_pairwise_sort_super_efficient_time.py
start_time = time.time()

n = int(input())                            # 1
max_number = float("-inf")                  # 1
second_max_number = float("-inf")           # 1

for num in input().split():                 # 3 + 2n

    num = int(num)                          # n

    if num > max_number:                    # n
        second_max_number = max_number      # [0, n] in the worst case it would be n times
        max_number = num                    # [0, n] in the worst case it would be n times
    elif num > second_max_number:           # n
        second_max_number = num             # [0, n] in the worst case it would be n times

print(f"n: {n} Product: {max_number * second_max_number}")

end_time = time.time()
total_time = end_time - start_time

# Calcular el tiempo estimado por instrucción
time_per_instruction = total_time / instruction_count

# Calcular el número de instrucciones por segundo
instructions_per_second = instruction_count / total_time

print("Tiempo estimado por instrucción:", time_per_instruction, "segundos")
print("Instrucciones por segundo:", instructions_per_second)

Writing max_pairwise_sort_super_efficient_time.py


### Super efficient solution in Python

Whitout storing the list in memory and whitout calling a function, add in each line a coment with the number of intstrcution that are executed in that line

Test with inputs 1 and 2




Test and time it with input 3

20 stress tests

Count executed instructions in code and verify with inputs 1,2 and 3

Calculate estimated time by instruction and number of instructions by second

### Estimate and verify the maximun n that can be solved in 1 second

## Perform 100 randomized automated tests in C++ and Python with a porgram  compares the answers,for the same 100 random input instances, of the super efficient solution $(T(n) < 10 n = O(n))$ against the sorting solution  $(T(n)  = O(n \lg{n}))$,  if the answers are not the same stops and displays "Error" listing  the input and the two answers; if the 100 tests are completed without stoping displays "Pass".

Super efficient solution (𝑇(𝑛)<10𝑛=𝑂(𝑛))

In [ ]:

%%writefile max_pairwise_sort_super_efficient.py
n = int(input())                            # 1
max_number = float("-inf")                  # 1
second_max_number = float("-inf")           # 1

for num in input().split():                 # 3 + 2n

    num = int(num)                          # n

    if num > max_number:                    # n
        second_max_number = max_number      # [0, n] in the worst case it would be n times
        max_number = num                    # [0, n] in the worst case it would be n times
    elif num > second_max_number:           # n
        second_max_number = num             # [0, n] in the worst case it would be n times

print(f"n: {n} Product: {max_number * second_max_number}")

Sortin solution (𝑇(𝑛)=𝑂(𝑛𝑙𝑛𝑔))

In [ ]:

%%writefile max_pairwise_sort_nlogn.py
# Read input values
n = int(input())              # 1
numbers = list(map(int, input().split()))  # n

# Sort the list
numbers.sort()                # n*log(n)

# Calculate the maximum pairwise product
result = numbers[-1] * numbers[-2]  # 1

# Print the result
print(f"n: {n} Product: {result}")